### ISSUE TITLE
Improve Graph Edge Visiblity

### ISSUE THREAD

**Issue by AGDCservices** [NationalSecurityAgency/ghidra#1406]
**Is your feature request related to a problem? Please describe.**
When viewing a binary in the graph view, I often come across graph edges that overlap which makes it very difficult to distinguish what is going on.

**Describe the solution you'd like**
I'd like to suggest updating the graphing algorithm to include a minimum distance between all edges so that you can clearly follow the edge without having to double click the line to see where it goes.

**Additional context**
An example screenshot is below.  Note that the green edges overlap coming out of the lower code block as well as the green edge overlapping with the incoming red edge at the upper code block.

![image](https://user-images.githubusercontent.com/57851463/71720399-9cd72300-2def-11ea-92ce-914ae342bf88.png)

**Comment by dragonmacher** [comment]
There has been work in this area recently.  

There is an option that changes the edge routing, which will partially separate the edges:

`Edit -> Tool Options... -> Function Graph -> Nested Code Layout -> Route Edges Around Vertices`


See if this helps for what you need.   We may need a different option to separate the edges regardless of the routing being used.

**Comment by AGDCservices** [comment]
Thanks for the tip.  I tried playing with different combinations of the "Route Edges Around Vertices" and "Use Condensed Layout" in case the condensed layout caused the overlapping, but the overlapping seems to occur with pretty much all of the combinations.  It seems a little better with the Route Edges Around Vertices, but i still get a fair amount of overlap.

**Comment by AGDCservices** [comment]
If you're looking at possibly adding different layout options, have you looked at Cutter (https://github.com/rizinorg/cutter)? Their graphing algorithm seems to work pretty well and is a lot easier to understand the code flow. It looks about the same as Ida Pro, but since it's open source, maybe you could add a similar algorithm. I think that would be a great addition to Ghidra. The decompiler works great, but the graphing definitely isn't quite as nice. If Ghidra had similar graphing to Ida Pro or Cutter, it would make it a lot easier to use for malware analysis. I look mostly at the assembly so I can get used to what assembly patterns are common vs unique for creating Yara rules, so i tend not to use the decomplier much at all and would love a graphing view that made it easier to understand the control flow.

**Comment by dragonmacher** [comment]
In a separate ticket we are looking at adding other layouts to the graph.    We will take a look at how cutter performs its layouts.    Having input like this for creating and updating layouts is extremely valuable.    What would you say are the traits that make the cutter or IDA layouts useful for you?

**Comment by AGDCservices** [comment]
Sorry for the delay. I've been trying to find some good examples but it's hard to show in a screen shot versus when you scroll through a binary. i put a few pictures to show roughly what i'm talking about. the pictures aren't overly bad for ghidra, just what i could find to highlight the points. when i was looking for examples, i found that the issue is you really have to scroll through the ghidra graph a good bit to really get an idea vs in cutter / ida you can more easily see everything with less scrolling. it's something that can be deomonstrated in person, but harder than i thought to clearly outline in a single picture.

When reviewing the code in the graph view, you're trying to easily follow the code logic without having to review all the conditional statements. After that, you're really looking to compare alternate code paths in the easiest manner to see why one would be chosen vs the other, etc.

Probably the few most helpful aspects for the graph view to support that are:
1. have the graph edges clearly separated and routed in a somewhat straight line when possible
2. having alternate code paths side by side to you can see what differences are between them
3. having consistent colors for all the graph edges

for point 1, this is where ghidra could use the most update. some of the previous updates make the graph edges better by routing around code blocks instead of going behind them, but the graph edges tend to overlap or are very close to each other which makes it hard to follow without double clicking etc. Additionally, ghidra routes the edges very close to the code blocks which tends to make them not go in straight lines because they are constantly routing around new code blocks. Cutter and Ida both keep a little more distance between edges and between the edges and the code blocks. They're not so far away to cause you to scroll over to see them, just far enough away so they keep straight lines much more frequently instead of having to route around new code blocks. this lets you roughly skim the code blocks and follow the edges without a lot of effort.

for point 2, ghidra puts if / else code blocks in a vertical column instead of side by side like Cutter and Ida Pro. This is partially just a preference, and with a small if / else it makes little difference. But when you have an if / else where each path has multiple follow on code blocks, it makes a big difference. when the if / else blocks are aligned vertically, you can't see the alternate paths next to each other and have to review one path, then review the second, and try ot remember what you saw to determine the differences, etc. When the paths are side by side, you can scroll down and see each path's code block next to each other and see what the difference is very easily.

point 3 is the smallest point. in cutter and ida, you only have 3 edge colors, red, green, and blue. All conditional jumps use red / green and absolute jumps use blue. ghidra does the same for conditional jumps, but the absolute jumps can be both blue and red, depending on if it's an actual jmp or if it's a fall through from one code block to another. while this is a minor point, when you're skimming, having clearly separate colors for conditional logic is really handy. when the fall through paths are red, it looks like a conditional jump unless you read the actual instruction. Because a large value in the graph view is to see the assembly instructions in a way that makes it easy to follow the logical flow without having to read very close, the over-use of the red edge makes this more difficult.

these first 2 pictures show the same location in ghidra / cutter. you can see the lines are clearly separated in cutter but overlapping in ghidra so you can't visually follow the lines. in cutter, you can see both paths of the if / else together and easily see the differences where the two different paths in ghidra are separated fairly far in the graph because of the vertical alignment so that you can't see them next to each other. 

![ghidra2](https://user-images.githubusercontent.com/57851463/123859250-6f8b0880-d8f2-11eb-8d3b-3eab85aa3288.PNG)


![cutter2](https://user-images.githubusercontent.com/57851463/123859268-731e8f80-d8f2-11eb-8e25-e895b05d1907.PNG)

these next 2 pictures show the same location in ghidra and cutter. you can see that in cutter, you get more code on the screen, you don't have overlapping lines where the jump out of the if / else crosses over the if / else input edge, etc. this example isn't bad in ghidra, but hopefully you can see it's a little more clear in cutter.

![ghidra1](https://user-images.githubusercontent.com/57851463/123859306-7c0f6100-d8f2-11eb-8dd5-a4ce92db0763.PNG)


![cutter1](https://user-images.githubusercontent.com/57851463/123859319-7f0a5180-d8f2-11eb-9f04-2ce20453c106.PNG)

**Comment by dragonmacher** [comment]
Thanks for the update.  This info is helpful.

Points 1 and 2 will likely require a new custom layout.   We are experimenting with adding other common graph layouts, but none of those are producing something similar to what you are describing.    

We should be able to fix the edge coloring issue fairly easily.

**Comment by sar5430** [comment]
I strongly agree with @AGDCservices on point 1 and 2. I feel like it's just easier and faster when if/else blocks are rendered in a vertical way.

Now that we have the "Graph" menu, I realize that the layout "Vertical Hierarchical MinCross Top Down" is certainly (at least for me) a way more readable graph layout. 

Here's a small example: 

Ghidra "classical" Nested layout: 

![Screenshot 2021-10-01 at 12 10 17](https://user-images.githubusercontent.com/15121293/135605084-047c53a3-90dd-42c4-b084-a66f2d131217.png)

Binary ninja layout: 
![Screenshot 2021-10-01 at 12 10 32](https://user-images.githubusercontent.com/15121293/135605119-7cf24443-9dfb-4a7b-9fce-b35e7d85e3b4.png)

Vertical Hierarchical MinCross Top Down "Graph" layout:
![Screenshot 2021-10-01 at 12 15 13](https://user-images.githubusercontent.com/15121293/135605171-f758e023-dbe0-4f0a-86d0-20c986971b3a.png)

I guess that adding `Vertical Hierarchical MinCross Top Down` as a possible layout for "Function Graph" could be a great addition. If I'm correct, both layout are part of Jung library ? So I guess it would not require to implement a totally new custom layout ?

**Comment by dragonmacher** [comment]
To this point, there are now a handful of these layouts available in the Function Graph in the master branch.    If you can build Ghidra, then you should give them a try.  Otherwise, they will appear in 10.1.

We may still need another layout tailored specifically to the Function Graph, similar to other tools.

**Comment by sar5430** [comment]
Great news! I'll try that as soon as I can. Thanks for the info
